pkg/server/tls/dynamic_source.go

```diff
@@ -118,44 +118,37 @@ func (f *DynamicSource) Start(ctx context.Context) error {
 
 	// channel which will be notified when the leaf certificate reaches 2/3 of its lifetime
 	// and needs to be renewed
-	renewalChan := make(chan struct{})
+	renewalChan := make(chan struct{}, 1)
 	group.Go(func() error {
-		// At this point, we expect to have one renewal moment
-		// in the channel, so we can start the timer with that value
 		var renewMoment time.Time
-		select {
-		case renewMoment = <-nextRenewCh:
-			// We recevieved a renew moment
-		default:
-			// This should never happen
-			panic("Unreacheable")
-		}
 
 		for {
 			if done := func() bool {
-				timer := time.NewTimer(time.Until(renewMoment))
-				defer timer.Stop()
+				var timerChannel <-chan time.Time
+				if !renewMoment.IsZero() {
+					timer := time.NewTimer(time.Until(renewMoment))
+					defer timer.Stop()
+
+					renewMoment = time.Time{}
+					timerChannel = timer.C
+				}
 
 				// Wait for the timer to expire, or for a new renewal moment to be received
 				select {
 				case <-ctx.Done():
 					// context was cancelled, return nil
 					return true
-				case <-timer.C:
+				case <-timerChannel:
 					// Continue to the next select to try to send a message on renewalChan
 				case renewMoment = <-nextRenewCh:
 					// We recevieved a renew moment, next loop iteration will update the timer
 					return false
 				}
 
-				// Try to send a message on renewalChan, but also allow for the context to be
```

Comment:
Instead of blocking on `renewalChan`, we send a message if the channel is not full yet similar to how `rotationChan` works.